File: src/main/scala/com/minalien/mffs/items/cards/ItemPowerLinkCard.scala
package com.minalien.mffs.items.cards



import com.minalien.mffs.api.EnumCardType.EnumCardType
import com.minalien.mffs.api.{IPowerLinkItem, EnumCardType, IForceStorageEnergyBlock}
import com.minalien.mffs.core.LinkGrid
import com.minalien.mffs.tiles.{MFFSMachine, TileEntityExtractor, TileEntityCapacitor}
import com.minalien.mffs.utils.Functions
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.{EnumFacing, BlockPos}
import net.minecraft.world.World

/**
 * Created by Katrina on 25/01/2015.
 */
object ItemPowerLinkCard extends ItemCard with IPowerLinkItem{
  var storage: IForceStorageEnergyBlock = null

  override def getCardType(): EnumCardType = EnumCardType.PowerLinkCard

  override def onUpdate(stack: ItemStack, worldIn: World, entityIn: Entity, itemSlot: Int, isSelected: Boolean): Unit = {
    super.onUpdate(stack, worldIn, entityIn, itemSlot, isSelected)
    if (Tick > 600) {
      val capID = this.getValuefromKey("CapacitorID", stack)
      if (capID != 0) {
        val cap = LinkGrid.getWorldNet(worldIn).Capacitors.lift(capID)
        if(!cap.isEmpty) {
          cap match {
            case Some(capacitor) => {
              if (!capacitor.getDeviceName.equals(this.getForArea(stack)))
                this.setForArea(stack, capacitor.getDeviceName)
            }
          }
        }

      }
      Tick = 0

    }
    Tick = Tick + 1
  }

  override def onItemUseFirst(stack: ItemStack, player: EntityPlayer, world: World, pos: BlockPos, side: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): Boolean = {
    val tileEntity: TileEntity = world.getTileEntity(pos)
    // TODO: Give this section security!
    if (!world.isRemote) {
      tileEntity match {
        case tile: TileEntityCapacitor => {
          Functions.setIteminSlot(stack, player, tile, 2, "<Power-Link>")
        }
        case tile: TileEntityExtractor => {
          if ((tileEntity.asInstanceOf[TileEntityExtractor]).getStackInSlot(1) == null) {
            (tileEntity.asInstanceOf[TileEntityExtractor]).setInventorySlotContents(1, stack)
            player.inventory.mainInventory(player.inventory.currentItem) = null
            Functions.chatToPlayer(player, "linkCard.installed")
            true
          } else {
            if ((tileEntity.asInstanceOf[TileEntityExtractor]).getStackInSlot(1).getItem eq ItemCardEmpty) {
              val itemstackcopy: ItemStack = stack.copy
              (tileEntity.asInstanceOf[TileEntityExtractor]).setInventorySlotContents(1, itemstackcopy)
              Functions.chatToPlayer(player, "linkCard.copied")
              true
            }
            else {

              Functions.chatToPlayer(player, "linkCard.notEmpty")
              false
            }
          }
        }
        case _ => false
      }
    }
    else
      false

  }


  def getForceEnergyStorageBlock(itemStack: ItemStack, tem: MFFSMachine, world: World): IForceStorageEnergyBlock = {
    if (itemStack != null && itemStack.getItem.isInstanceOf[ItemCard]) {
      if (itemStack.getItem.asInstanceOf[ItemCard].isValid(itemStack)) {
        val png = this.getCardTargetPoint(itemStack)
        if (png != null) {
          if (png.dimID != world.provider.getDimensionId)
            return null
          if (world.getTileEntity(png.pos).isInstanceOf[TileEntityCapacitor]) {
            val cap = world.getTileEntity(png.pos).asInstanceOf[TileEntityCapacitor]
            if (cap != null) {
              if (cap.getPowerStorageID == this.getValuefromKey("CapacitorID", itemStack) && this.getValuefromKey("CapacitorID", itemStack) != 0) {
                if (!cap.getDeviceName.equals(this.getForArea(itemStack)))
                  this.setForArea(itemStack, cap.getDeviceName)

                if ((cap.getTransmitRange * cap.getTransmitRange) >= tem.getMachinePoint.distanceSq(cap.getMachinePoint))
                  return cap
              }
              return null
            }
          }
          else {
            val CapID = this.getValuefromKey("CapacitorID", itemStack)
            if (CapID != 0) {
              val cap = LinkGrid.getWorldNet(world).Capacitors.lift(CapID)
              cap match {
                case Some(capacitor) => {
                  this.setInformation(itemStack, capacitor.getMachinePoint, "CapacitorID", CapID)
                  if ((capacitor.getTransmitRange*capacitor.getTransmitRange) >= tem.getMachinePoint.distanceSq(capacitor.getMachinePoint))
                    return capacitor
                  return null
                }
              }
            }
          }
          if (world.getChunkFromBlockCoords(png.pos).isLoaded)
            this.setInvalid(itemStack)

        }
      }
    }
    return null
  }

  override def getPercentageCapacity(itemStack: ItemStack, pos: BlockPos, world: World): Int =
  {
    storage=getForceEnergyStorageBlock(itemStack,world.getTileEntity(pos).asInstanceOf[MFFSMachine],world)
    if(storage!=null)
      storage.getPercentageStorageCapacity
    else
      0
  }

  override def insertPower(itemStack: ItemStack, powerAmount: Int, simulation: Boolean, pos: BlockPos, world: World): Boolean =
  {
    storage=getForceEnergyStorageBlock(itemStack,world.getTileEntity(pos).asInstanceOf[MFFSMachine],world)
    if(storage!=null)
      storage.insertPowertoStorage(powerAmount,simulation)
    else
      false
  }

  override def getAvailablePower(itemStack: ItemStack, pos: BlockPos, world: World): Int =
  {
    storage=getForceEnergyStorageBlock(itemStack,world.getTileEntity(pos).asInstanceOf[MFFSMachine],world)
    if(storage!=null)
      storage.getStorageAvailablePower
    else
      0
  }

  override def consumePower(itemStack: ItemStack, powerAmount: Int, simulation: Boolean, pos: BlockPos, world: World): Boolean =
  {
    storage=getForceEnergyStorageBlock(itemStack,world.getTileEntity(pos).asInstanceOf[MFFSMachine],world)
    if(storage!=null)
      storage.consumePowerfromStorage(powerAmount,simulation)
    else
      false
  }

  override def getfreeStorageAmount(itemStack: ItemStack, pos: BlockPos, world: World): Int =

  {
    storage=getForceEnergyStorageBlock(itemStack,world.getTileEntity(pos).asInstanceOf[MFFSMachine],world)
    if(storage!=null)
      storage.getfreeStorageAmount
    else
      0
  }

  override def getPowersourceID(itemStack: ItemStack, pos: BlockPos, world: World): Int =
  {
    storage=getForceEnergyStorageBlock(itemStack,world.getTileEntity(pos).asInstanceOf[MFFSMachine],world)
    if(storage!=null)
      storage.getPowerStorageID
    else
      0
  }

  override def isPowersourceItem: Boolean = false

  override def getMaximumPower(itemStack: ItemStack, pos: BlockPos, world: World): Int =
  {
    storage=getForceEnergyStorageBlock(itemStack,world.getTileEntity(pos).asInstanceOf[MFFSMachine],world)
    if(storage!=null)
      storage.getStorageMaxPower
    else
      1
  }
}

File: src/main/java/net/newgaea/mffs/common/items/linkcards/ItemPowerLinkCard.java
package net.newgaea.mffs.common.items.linkcards;

import net.minecraft.entity.Entity;
import net.minecraft.entity.item.ItemEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ItemUseContext;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.ActionResultType;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.ChunkPos;
import net.minecraft.world.World;
import net.newgaea.mffs.api.IForceEnergyStorageBlock;
import net.newgaea.mffs.api.IPowerLinkItem;
import net.newgaea.mffs.common.data.Grid;
import net.newgaea.mffs.common.tiles.TileCapacitor;
import net.newgaea.mffs.common.tiles.TileNetwork;

public class ItemPowerLinkCard extends ItemCard  implements IPowerLinkItem {


    public ItemPowerLinkCard(Properties properties) {
        super(properties);
    }

    @Override
    public void inventoryTick(ItemStack stack, World worldIn, Entity entity, int itemSlot, boolean isSelected) {
        super.inventoryTick(stack, worldIn, entity, itemSlot, isSelected);
        if(this.Tick > 600) {
            int cap_id=this.getValueFromKey("capacitorID",stack);
            if(cap_id!=0) {
                TileCapacitor cap=Grid.getWorldGrid(entity.getEntityWorld()).getCapacitors().get(cap_id);
                if(cap!=null) {
                    if(!cap.getDeviceName().equals(this.getAreaName(stack)))
                        this.setAreaName(stack,cap.getDeviceName());
                }
            }
            Tick=0;
        }
        Tick++;
    }

    @Override
    public boolean onEntityItemUpdate(ItemStack stack, ItemEntity entity) {

        return super.onEntityItemUpdate(stack, entity);
    }


    @Override
    public ActionResultType onItemUseFirst(ItemStack stack, ItemUseContext context) {

        if(!context.getWorld().isRemote) {
            TileEntity tileEntity = context.getWorld().getTileEntity(context.getPos());
            // todo: auto insert
        }
        return ActionResultType.PASS;
    }

    public IForceEnergyStorageBlock getForceEnergyStorageBlock(ItemStack stack, TileNetwork tem, World world) {
        if(!stack.isEmpty() && stack.getItem() instanceof ItemCard) {
            if(((ItemCard)stack.getItem()).isValid(stack)) {
                BlockPos pos=this.getCardTargetPoint(stack);
                if(pos!=null) {
                    if(world.getTileEntity(pos) instanceof TileCapacitor) {
                        TileCapacitor cap= (TileCapacitor) world.getTileEntity(pos);
                        int cap_id=this.getValueFromKey("capacitorID",stack);
                        if(cap.getPowerStorageID() == cap_id && cap_id!=0) {
                            if(!cap.getDeviceName().equals(ItemCard.getAreaName(stack))) {
                                ItemCard.setAreaName(stack,cap.getDeviceName());
                            }
                            if(Math.pow(cap.getTransmitRange(),2) >= tem.getPos().distanceSq(cap.getPos()))
                                return cap;
                            return null;
                        }
                    }
                    else {
                        int cap_id=this.getValueFromKey("capacitorID",stack);
                        if(cap_id!=0) {
                            TileCapacitor cap=Grid.getWorldGrid(world).getCapacitors().get(cap_id);
                            if(cap!=null) {
                                ((ItemCard)this).setInformation(stack,cap.getPos(),"capacitorID",cap_id,world);
                                if(Math.pow(cap.getTransmitRange(),2) >= tem.getPos().distanceSq(cap.getPos()))
                                    return cap;
                                return null;
                            }
                        }
                    }
                    if(world.getChunkProvider().isChunkLoaded(new ChunkPos(pos)))
                        ((ItemCard)stack.getItem()).setInvalid(stack);
                }
            }
        }
        return null;
    }

    @Override
    public int getAvailablePower(ItemStack stack, TileNetwork machine, World world) {
        IForceEnergyStorageBlock storage=getForceEnergyStorageBlock(stack,machine,world);
        if(storage!=null)
            return storage.getStorageAvailablePower();
        return 0;
    }

    @Override
    public int getMaximumPower(ItemStack stack, TileNetwork machine, World world) {
        IForceEnergyStorageBlock storage=getForceEnergyStorageBlock(stack,machine,world);
        if(storage!=null)
            return storage.getStorageMaxPower();
        return 0;
    }
    @Override
    public int getPowersourceID(ItemStack stack, TileNetwork machine, World world) {
        IForceEnergyStorageBlock storage=getForceEnergyStorageBlock(stack,machine,world);
        if(storage!=null)
            return storage.getPowerStorageID();
        return 0;
    }

    @Override
    public int getPercentageCapacity(ItemStack stack, TileNetwork machine, World world) {
        IForceEnergyStorageBlock storage=getForceEnergyStorageBlock(stack,machine,world);
        if(storage!=null)
            return storage.getPercentageStorageCapacity();
        return 0;
    }




    @Override
    public boolean consumePower(ItemStack stack, int powerAmount, boolean simulation, TileNetwork machine, World world) {
        IForceEnergyStorageBlock storage=getForceEnergyStorageBlock(stack,machine,world);
        if(storage!=null)
            return storage.consumePowerfromStorage(powerAmount, simulation);
        return false;
    }

    @Override
    public boolean insertPower(ItemStack stack, int powerAmount, boolean simulation, TileNetwork machine, World world) {
        IForceEnergyStorageBlock storage=getForceEnergyStorageBlock(stack,machine,world);
        if(storage!=null)
            return storage.insertPowertoStorage(powerAmount, simulation);
        return false;
    }


    @Override
    public int getFreeStorageAmount(ItemStack stack, TileNetwork machine, World world) {
        IForceEnergyStorageBlock storage=getForceEnergyStorageBlock(stack,machine,world);
        if(storage!=null)
            return storage.getfreeStorageAmount();
        return 0;
    }

    @Override
    public boolean isPowerSourceItem() {
        return false;
    }


}
